### PR TITLE
Extend `isTupleLikeType` to also check if `.length` is a number literal type

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/29311

implements the suggested solution by @ahejlsberg from [here](https://github.com/microsoft/TypeScript/issues/29311#issuecomment-452780655)

### CODE REVIEW

**Comment by Andarist** [comment]
@jakebailey I wonder if this doesn't also resolve what you were trying to solve [here](https://github.com/microsoft/TypeScript/pull/52551), would you mind taking a look?

**Comment by jakebailey** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test tsserver top100
@typescript-bot run dt
@typescript-bot perf test this
@typescript-bot pack this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -22914,7 +22914,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function isTupleLikeType(type: Type): boolean {
-        return isTupleType(type) || !!getPropertyOfType(type, "0" as __String);
+        let lengthType;
+        return isTupleType(type) ||
+            !!getPropertyOfType(type, "0" as __String) ||
```
Comment (jakebailey) [inline src/compiler/checker.ts:22919]: I wonder how much it'd take to eliminate this check; shouldn't all tuples have a length property?

Comment (Andarist) [reply src/compiler/checker.ts:22919]: Do you mean the "0" prop check?

Comment (jakebailey) [reply src/compiler/checker.ts:22919]: Yes, though, judging by the baseline changes there are some other things that would need to be fixed, so maybe it's unwise to try.

Comment (Andarist) [reply src/compiler/checker.ts:22919]: Yeah, you have things like this in the test suite:
https://github.com/microsoft/TypeScript/blob/1ed0a5ac4b49971e4698d5388f7760cbaa0615ac/tests/cases/compiler/inferringAnyFunctionType1.ts

I think I might have seen `0` prop being added to things in the wild to hint at tuple inference and stuff. Arguably, they were abusing this particular heuristic a little bit - but it's also not like it was hidden (I've seen it mentioned on a couple of occasions, including in the linked ticket [here](https://github.com/microsoft/TypeScript/issues/29311#issuecomment-453337774))

Comment (Andarist) [reply src/compiler/checker.ts:22919]: @jakebailey any further thoughts about this PR? I've recently learned that `'0'` thing was often~ used to imitate tuples when tuples were supported by TS

Comment (jakebailey) [reply src/compiler/checker.ts:22919]: Getting to this later than I intended; I don't think I have any remaining concerns here, though the code flow of `lengthType` is a little funky compared to a plain if statement. Not sure it matters. I bet there's some silly v8 thing where the expression is faster for some silly reason.

**Comment by jakebailey** [comment]
Paranoia check

@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this faster

**Comment by jakebailey** [comment]
I'm guessing that perf result is a lie caused by the main baseline being behind. Will rerun later.

**Comment by jakebailey** [comment]
@typescript-bot perf test this faster

File: src/compiler/checker.ts (hunk 1 above)
Comment (Andarist) [reply src/compiler/checker.ts:22919]: Hm, i wouldnt thought about this being faster (maybe it is, i dont know). It’s not exactly how i would write it - it’s just using a similar code style that is used in other places in the codebase.

**Comment by jakebailey** [comment]
@typescript-bot perf test this

**Review by jakebailey — APPROVED** [review]
Coming back to this, I think this is good (especially given I closed my own #52551 saying this PR is the correct one).

**Pull request merged**